smart_open/smart_open_lib.py

```diff
@@ -928,3 +928,28 @@ def _encoding_wrapper(fileobj, mode, encoding=None, errors=None):
     if mode[0] in ('w', 'a') or mode.endswith('+'):
         fileobj = codecs.getwriter(encoding)(fileobj, **kw)
     return fileobj
+
+
+class patch_pathlib(object):
+    """Replace `Path.open` with `smart_open.open`"""
+
+    def __init__(self):
+        self.old_impl = _patch_pathlib(open)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        _patch_pathlib(self.old_impl)
+
+
+def _patch_pathlib(func):
+    """Replace `Path.open` with `func`"""
+    pathlib = sys.modules.get("pathlib", None)
+
+    if not pathlib:
+        raise RuntimeError("Can't patch 'pathlib.Path.open', you should import 'pathlib' first")
```

Comment:
Why not import it ourselves?

Reply:
We shouldn't do that ourselves, this is bad tone.

BTW - this doesn't help to the user, because the user will not have `Path` class available after our import

Reply:
I don't follow – imported modules are singletons, it doesn't matter who imported it.

Reply:
that's true, but user has no access to `pathlib` itself (without digging into `sys.modules`), like 

```python
In [1]: def f(): 
   ...:     import pathlib 
   ...:                                                                         

In [2]: pathlib                                                                 
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-2-7bc182bbe5df> in <module>
----> 1 pathlib

NameError: name 'pathlib' is not defined
```

I expect that user UNDERSTAND what they do, for this reason, I expect than user firstly import `pathlib` and after apply monkeypathing.

Reply:
I still don't get it. What's the point of failing with "import pathlib first!", when we can import it ourselves?

Reply:
…plus `pathlib` already seems to be pre-imported internally:

```python
>>> import sys
>>> sys.modules.get("pathlib", None)
<module 'pathlib' from '/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pathlib.py'>
```

So when can this `if` condition actually happen? Is it some Python version compatibility thing?
If so, needs better comments (and a different exception message).

Reply:
In my environment:

```
(smart_open) sergeyich:smart_open misha$ python
Python 3.7.6 (default, Dec 30 2019, 19:38:28)
[Clang 11.0.0 (clang-1100.0.33.16)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.modules.get('pathlib', 'not there')
'not there'
```

However, `smart_open` already imports pathlib by itself: https://github.com/RaRe-Technologies/smart_open/blob/master/smart_open/smart_open_lib.py#L51

So, given that we're already doing it, it's probably simpler to patch it directly.

Reply:
Ok @mpenkov, if we do that anyway, I agree.